Do not define AVIF_ENABLE_NODISCARD for avif.h

wantehchang opened the pull request:
```
Make the AVIF_ENABLE_NODISCARD cmake option require C23 so that avif.h
can define the AVIF_NODISCARD macro as [[nodiscard]] without checking
the AVIF_ENABLE_NODISCARD macro. This requires changing avif.h to test
__STDC_VERSION__ >= 202000L, which is the value of __STDC_VERSION__ in
recent versions of GCC and Clang when -std=c2x or -std=gnu2x is
specified.

In general avif.h should only test compiler predefined macros. The only
exception is AVIF_DLL (ignoring the experimental feature macros.)

Fix https://github.com/AOMediaCodec/libavif/issues/2352.
```

wantehchang commented on CMakeLists.txt:
```diff
@@ -33,8 +33,16 @@ include(FetchContent)
 include(FindPkgConfig)
 include(AvifExternalProjectUtils)
 
+option(AVIF_ENABLE_NODISCARD "Add [[nodiscard]] to some functions. CMake must be at least 3.21 to request C23." OFF)
```
This is copied from line 55 in the original code, with "force C23" changed to "request C23."

wantehchang commented on CMakeLists.txt:
```diff
@@ -333,22 +340,6 @@ endif()
-    if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.21.0)
-        set(CMAKE_C_STANDARD 23)
-        set_property(TARGET avif_obj PROPERTY C_STANDARD 23)
-    else()
```
If we redefine `AVIF_ENABLE_NODISCARD` to be a request rather than a requirement, we don't need this `else` block.

wantehchang commented on CMakeLists.txt:
```diff
@@ -333,22 +340,6 @@ endif()
-            target_compile_options(avif_obj PUBLIC $<BUILD_INTERFACE:$<$<COMPILE_LANGUAGE:C>:/std:clatest>>)
-        endif()
-    endif()
-    target_compile_definitions(avif_obj PUBLIC $<BUILD_INTERFACE:AVIF_ENABLE_NODISCARD=1>)
```
The definition of the `AVIF_ENABLE_NODISCARD` macro for avif.h is also deleted.

wantehchang commented:
@vrabaud FYI.

jzern reviewed (approved):
This makes sense.

The decaying is unfortunate:

https://cmake.org/cmake/help/latest/manual/cmake-compile-features.7.html#id8

But it looks like [`CMAKE_C_STANDARD_REQUIRED`](https://cmake.org/cmake/help/latest/variable/CMAKE_C_STANDARD_REQUIRED.html) might work if we wanted to abort when the option was specified, but the version not available.

wantehchang commented:
> But it looks like [`CMAKE_C_STANDARD_REQUIRED`](https://cmake.org/cmake/help/latest/variable/CMAKE_C_STANDARD_REQUIRED.html) might work if we wanted to abort when the option was specified, but the version not available.

Thank you for the suggestion. I implemented this in the second commit. I also rewrote the commit message and pasted it at the top of this issue. Please take a look. Thanks!

jzern commented on CMakeLists.txt:
```diff
@@ -33,8 +33,19 @@ include(FetchContent)
+    set(CMAKE_C_STANDARD 23)
+    set(CMAKE_C_STANDARD_REQUIRED ON)
+else()
+    set(CMAKE_C_STANDARD 99)
```
I would think this would be required too, but it's highly unlikely anyone would try this with a c90 only compiler (or force those flags)

wantehchang replied:
Yes, we assume the compiler supports C99.
